### ISSUE TITLE
Unable to use okta_group data source as group_id for okta_app_group_assignment if not known at plan time

### ISSUE THREAD

**Issue by MatthewJohn** [okta/terraform-provider-okta#1642]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
Terraform v1.4.6
on linux_amd64
+ provider registry.terraform.io/okta/okta v3.46.0
```

Update: I've realised that our actual implementation is using the constraint "~> 4.0"

### Affected Resource(s)

- okta_app_group_assignment
- data.okta_group (possibly)

### Terraform Configuration Files

https://github.com/MatthewJohn/example-okta-group-data-issue/blob/main/main.tf
```
# Reproduction using example resources from documentation
resource "okta_app_saml" "example" {
  label                    = "tf-example"
  sso_url                  = "https://example.com"
  recipient                = "https://example.com"
  destination              = "https://example.com"
  audience                 = "https://example.com/audience"
  subject_name_id_template = "$${user.userName}"
  subject_name_id_format   = "urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress"
  response_signed          = true
  signature_algorithm      = "RSA_SHA256"
  digest_algorithm         = "SHA256"
  honor_force_authn        = false
  authn_context_class_ref  = "urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport"

  attribute_statements {
    type         = "GROUP"
    name         = "groups"
    filter_type  = "REGEX"
    filter_value = ".*"
  }
}

resource "okta_group" "example" {
  name        = "tf-example"
  description = "My Example Group"
}

data "okta_group" "example" {
  name = "tf-example"
  depends_on = [okta_group.example]
}

resource "okta_app_group_assignment" "example" {
  app_id   = okta_app_saml.example.id
  group_id = data.okta_group.example.id
}
```

### Debug Output

See https://github.com/MatthewJohn/example-okta-group-data-issue/blob/main/debug_output.log

### Panic Output

None

### Expected Behavior

The group_id from the data source is "known at apply" and the okta_app_group_assignment contains a group_id that is "known at apply".
The plan succeeds and the apply creates the Okta app and the Okta group. The data.okta_group then obtains the details of the okta group and the group assignment is created.

Backstory is what is trying to be performed - our groups are obtaining syncing from an external directory.
We wish to run terraform that creates the necessary groups (which creates the groups and only completes once the group is available in Okta).

As a result of creating the group outside of Okta, we do not initially have the group ID to be able to perform an okta_app_group_assignment, so we use a data source, which we use values of the group name that are only available at apply time, from the group creation process, so it is not attempted until we _know_ that the group will be available in Okta. The group ID from this data source is then passed to the group assignment.

The attached terraform appears to be able to demonstrate the issue.

#### Can this be done in the Admin UI?

Yes - but this is a terraform attribute issue.

#### Can this be done in the actual API call?

Yes - but this is a terraform attribute issue.

### Actual Behavior

An error is produced (see output)

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

N/A - the provided terraform is sufficient to reproduce the issue.

### References

- #0000

**Comment by MatthewJohn** [comment]
So, I've found _something_.. I suspect it's related to the fact that "id" of the data.okta_group resource is "optional" (which I know has been set to due to the input attribute being optional). However, I suspect Terraform is treating the attribute, entirely, as being optional (included the resulting attribute).

If I modify the `id` attribute to be `Computed`, the issue is resolved:
```
diff --git a/okta/data_source_okta_group.go b/okta/data_source_okta_group.go
index f6220db7..55502e74 100644
--- a/okta/data_source_okta_group.go
+++ b/okta/data_source_okta_group.go
@@ -17,9 +17,8 @@ func dataSourceGroup() *schema.Resource {
                ReadContext: dataSourceGroupRead,
                Schema: map[string]*schema.Schema{
                        "id": {
-                               Type:          schema.TypeString,
-                               Optional:      true,
-                               ConflictsWith: []string{"name", "type"},
+                               Type:     schema.TypeString,
+                               Computed: true,
                        },
                        "name": {
                                Type:     schema.TypeString,
```

I don't know enough about the Terraform provider - whether different attribute configurations can be provided for "input" (provided by the user) and "resulting" (generated by the data source) attributes.
Having said that, I _seem_ to recall a provider having two attributes for a single value (for example `id` and another attributed with a prefix.. I'm assuming that maybe one is used as the input and is set to `optional` and the other is the computed value (perhaps for this use case?).

So one "less clean" suggestion could be to add an additional property for the resolved ID of the object and one for the "user-provided" attribute?

e.g.
```
diff --git a/okta/data_source_okta_group.go b/okta/data_source_okta_group.go
index f6220db7..d7c71149 100644
--- a/okta/data_source_okta_group.go
+++ b/okta/data_source_okta_group.go
@@ -21,6 +21,10 @@ func dataSourceGroup() *schema.Resource {
                                Optional:      true,
                                ConflictsWith: []string{"name", "type"},
                        },
+                       "computed_id": {
+                               Type:     schema.TypeString,
+                               Computed: true,
+                       },
                        "name": {
                                Type:     schema.TypeString,
                                Optional: true,

```

**Comment by MatthewJohn** [comment]
Having said this - after testing, I can add "Computed" to true to the id resource and:
 * It fixes my use case
 * Data resources can still be obtained by name
 * Data resources can still be obtained by id

Maybe this would be a suitable fix? If so, maybe it could (or should, assuming it's a valid solution) be applied to the name attribute as well?

```
diff --git a/okta/data_source_okta_group.go b/okta/data_source_okta_group.go
index f6220db7..fb38e675 100644
--- a/okta/data_source_okta_group.go
+++ b/okta/data_source_okta_group.go
@@ -20,6 +20,7 @@ func dataSourceGroup() *schema.Resource {
                                Type:          schema.TypeString,
                                Optional:      true,
                                ConflictsWith: []string{"name", "type"},
+                               Computed:      true,
                        },
                        "name": {
                                Type:     schema.TypeString,
```

**Comment by duytiennguyen-okta** [comment]
OKTA internal reference https://oktainc.atlassian.net/browse/OKTA-631375

**Comment by duytiennguyen-okta** [comment]
This is fixed in the release [4.2.0](https://github.com/okta/terraform-provider-okta/releases/tag/v4.2.0) from https://github.com/okta/terraform-provider-okta/pull/1665
